File: TRAUMA-Python/license.py
#!/usr/bin/env python3
"""
TRAUMA Unified License System
One license works across all TRAUMA tools: OSINT, Python, Browser, Web
"""

import os
import sys
import json
import hmac
import hashlib
from datetime import datetime, timedelta
from pathlib import Path

# Shared secret - SAME across all TRAUMA products
TRAUMA_SECRET = 'TRUMA-OSINT-SECRET-2024-SECURE-KEY-DO-NOT-SHARE'

def get_app_dir():
    """Get the application directory"""
    if getattr(sys, 'frozen', False):
        # Running as compiled exe
        return Path(sys.executable).parent
    else:
        return Path(__file__).parent

def get_license_file():
    """Get the license file path"""
    return get_app_dir() / 'license.key'

def get_licenses_dir():
    """Get the licenses directory"""
    lic_dir = get_app_dir() / 'licenses'
    lic_dir.mkdir(exist_ok=True)
    return lic_dir

def generate_license(user='User', expiry_days=365):
    """Generate a new TRAUMA license key"""
    timestamp = int(datetime.now().timestamp() * 1000)
    expiry = timestamp + (expiry_days * 24 * 60 * 60 * 1000)
    payload = f"{user}|{timestamp}|{expiry}"
    
    signature = hmac.new(
        TRAUMA_SECRET.encode(),
        payload.encode(),
        hashlib.sha256
    ).hexdigest()[:32].upper()
    
    key = f"TRUMA-{signature[0:4]}-{signature[4:8]}-{signature[8:12]}-{signature[12:16]}-{signature[16:20]}-{signature[20:24]}"
    
    license_data = {
        'key': key,
        'user': user,
        'created': datetime.fromtimestamp(timestamp/1000).isoformat(),
        'expires': datetime.fromtimestamp(expiry/1000).isoformat(),
        'payload': payload
    }
    
    # Save to licenses directory
    lic_file = get_licenses_dir() / f"{user}_{timestamp}.json"
    with open(lic_file, 'w') as f:
        json.dump(license_data, f, indent=2)
    
    return license_data

def verify_license_format(key):
    """Verify license key format"""
    if not key or not isinstance(key, str):
        return False
    
    key = key.upper().strip()
    parts = key.split('-')
    
    if len(parts) != 7:
        return False
    
    if parts[0] != 'TRUMA':
        return False
    
    # Check each group is 4 alphanumeric chars
    for i in range(1, 7):
        if len(parts[i]) != 4:
            return False
        if not parts[i].isalnum():
            return False
    
    return True

def check_license():
    """Check if a valid license is activated"""
    lic_file = get_license_file()
    
    if not lic_file.exists():
        return {
            'valid': False,
            'error': 'No license activated',
            'needs_activation': True
        }
    
    try:
        with open(lic_file, 'r') as f:
            data = json.load(f)
        
        if not verify_license_format(data.get('key', '')):
            return {
                'valid': False,
                'error': 'Invalid license format'
            }
        
        # Check expiry
        if 'expires' in data:
            expiry = datetime.fromisoformat(data['expires'].replace('Z', '+00:00').replace('+00:00', ''))
            if expiry < datetime.now():
                return {
                    'valid': False,
                    'error': 'License expired',
                    'expired': True,
                    'expires': data['expires']
                }
        
        return {
            'valid': True,
            'user': data.get('user', 'Licensed User'),
            'expires': data.get('expires'),
            'created': data.get('created')
        }
    except Exception as e:
        return {
            'valid': False,
            'error': f'License file error: {str(e)}'
        }

def activate_license(key):
    """Activate a license key"""
    if not verify_license_format(key):
        return {
            'success': False,
            'error': 'Invalid license format'
        }
    
    # Create license data
    now = datetime.now()
    expiry = now + timedelta(days=365)
    
    license_data = {
        'key': key.upper(),
        'user': 'Licensed User',
        'created': now.isoformat(),
        'expires': expiry.isoformat(),
        'activated': int(now.timestamp() * 1000)
    }
    
    lic_file = get_license_file()
    with open(lic_file, 'w') as f:
        json.dump(license_data, f, indent=2)
    
    return {
        'success': True,
        'message': 'License activated successfully',
        'user': license_data['user'],
        'expires': license_data['expires']
    }

def deactivate_license():
    """Deactivate/remove license"""
    lic_file = get_license_file()
    if lic_file.exists():
        lic_file.unlink()
        return {'success': True, 'message': 'License deactivated'}
    return {'success': False, 'error': 'No license to deactivate'}

def get_license_info():
    """Get license info for display"""
    lic_file = get_license_file()
    
    if not lic_file.exists():
        return None
    
    try:
        with open(lic_file, 'r') as f:
            return json.load(f)
    except:
        return None

# CLI interface
if __name__ == '__main__':
    import argparse
    
    parser = argparse.ArgumentParser(description='TRAUMA Unified License System')
    parser.add_argument('--generate', '-g', action='store_true', help='Generate a new license')
    parser.add_argument('--user', '-u', type=str, default='User', help='Username for license')
    parser.add_argument('--days', '-d', type=int, default=365, help='Expiry days')
    parser.add_argument('--check', '-c', action='store_true', help='Check license status')
    parser.add_argument('--activate', '-a', type=str, help='Activate a license key')
    parser.add_argument('--deactivate', action='store_true', help='Deactivate license')
    
    args = parser.parse_args()
    
    if args.generate:
        lic = generate_license(args.user, args.days)
        print(f"\n{'='*60}")
        print(f"  TRAUMA UNIFIED LICENSE GENERATOR")
        print(f"{'='*60}")
        print(f"  User: {lic['user']}")
        print(f"  Created: {lic['created']}")
        print(f"  Expires: {lic['expires']}")
        print(f"{'='*60}")
        print(f"  KEY: {lic['key']}")
        print(f"{'='*60}\n")
    
    elif args.check:
        result = check_license()
        print(json.dumps(result, indent=2))
        sys.exit(0 if result['valid'] else 1)
    
    elif args.activate:
        result = activate_license(args.activate)
        print(json.dumps(result, indent=2))
        sys.exit(0 if result['success'] else 1)
    
    elif args.deactivate:
        result = deactivate_license()
        print(json.dumps(result, indent=2))
    
    else:
        parser.print_help()

File: TRUMA-OSINT/license.js
#!/usr/bin/env node

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// Get correct directory for both dev and pkg EXE
const APP_DIR = process.pkg ? path.dirname(process.execPath) : __dirname;

// License configuration
const LICENSES_DIR = path.join(APP_DIR, 'licenses');
const LICENSE_FILE = path.join(APP_DIR, 'license.key');
const SECRET_KEY = 'TRUMA-OSINT-SECRET-2024-SECURE-KEY-DO-NOT-SHARE';

// Ensure licenses directory exists
if (!fs.existsSync(LICENSES_DIR)) {
  fs.mkdirSync(LICENSES_DIR, { recursive: true });
}

function generateLicenseKey(userInfo = 'User', expiryDays = 365) {
  const timestamp = Date.now();
  const expiry = timestamp + (expiryDays * 24 * 60 * 60 * 1000);
  const payload = `${userInfo}|${timestamp}|${expiry}`;
  
  const signature = crypto
    .createHmac('sha256', SECRET_KEY)
    .update(payload)
    .digest('hex')
    .substring(0, 32)
    .toUpperCase();
  
  const key = `TRUMA-${signature.substring(0, 4)}-${signature.substring(4, 8)}-${signature.substring(8, 12)}-${signature.substring(12, 16)}-${signature.substring(16, 20)}-${signature.substring(20, 24)}`;
  
  return {
    key,
    payload,
    created: new Date(timestamp).toISOString(),
    expires: new Date(expiry).toISOString(),
    user: userInfo
  };
}

function verifyLicenseKey(key) {
  if (!key || typeof key !== 'string') return { valid: false, error: 'No key provided' };
  
  const parts = key.split('-');
  if (parts.length !== 7 || parts[0] !== 'TRUMA') {
    return { valid: false, error: 'Invalid key format' };
  }
  
  const signature = parts.slice(1).join('');
  if (signature.length !== 24) {
    return { valid: false, error: 'Invalid signature length' };
  }
  
  // Check if license file exists with stored payload
  if (fs.existsSync(LICENSE_FILE)) {
    try {
      const data = JSON.parse(fs.readFileSync(LICENSE_FILE, 'utf8'));
      const expectedSig = crypto
        .createHmac('sha256', SECRET_KEY)
        .update(data.payload)
        .digest('hex')
        .substring(0, 32)
        .toUpperCase();
      
      if (signature === expectedSig.substring(0, 24)) {
        const expiry = new Date(data.expires);
        if (expiry < new Date()) {
          return { valid: false, error: 'License expired', expired: true };
        }
        return { 
          valid: true, 
          user: data.user,
          expires: data.expires,
          created: data.created
        };
      }
    } catch (e) {}
  }
  
  // For newly entered keys, verify format only (full verification needs payload)
  if (/^[A-Z0-9]{4}$/.test(parts[1]) && 
      /^[A-Z0-9]{4}$/.test(parts[2]) &&
      /^[A-Z0-9]{4}$/.test(parts[3]) &&
      /^[A-Z0-9]{4}$/.test(parts[4]) &&
      /^[A-Z0-9]{4}$/.test(parts[5]) &&
      /^[A-Z0-9]{4}$/.test(parts[6])) {
    return { valid: true, provisional: true };
  }
  
  return { valid: false, error: 'Invalid key signature' };
}

function saveLicense(key, payload) {
  fs.writeFileSync(LICENSE_FILE, JSON.stringify({ key, payload, verified: Date.now() }, null, 2));
}

function activateLicense(key) {
  const result = verifyLicenseKey(key);
  if (result.valid && result.provisional) {
    // Create new payload for new activation
    const timestamp = Date.now();
    const expiry = timestamp + (365 * 24 * 60 * 60 * 1000);
    const payload = `Activated|${timestamp}|${expiry}`;
    saveLicense(key, { key, payload, created: new Date(timestamp).toISOString(), expires: new Date(expiry).toISOString(), user: 'Activated' });
    return { success: true, message: 'License activated successfully' };
  }
  return { success: false, error: result.error };
}

function checkLicense() {
  if (!fs.existsSync(LICENSE_FILE)) {
    return { valid: false, error: 'No license file found' };
  }
  
  try {
    const data = JSON.parse(fs.readFileSync(LICENSE_FILE, 'utf8'));
    return verifyLicenseKey(data.key);
  } catch (e) {
    return { valid: false, error: 'Corrupted license file' };
  }
}

// CLI Interface for key generation (owner only)
if (process.argv.includes('--generate')) {
  const user = process.argv[process.argv.indexOf('--generate') + 1] || 'User';
  const days = parseInt(process.argv[process.argv.indexOf('--days') + 1]) || 365;
  
  const license = generateLicenseKey(user, days);
  
  console.log('\n╔════════════════════════════════════════════════════════════╗');
  console.log('║              TRUMA OSINT LICENSE GENERATOR                 ║');
  console.log('╠════════════════════════════════════════════════════════════╣');
  console.log('║                                                            ║');
  console.log(`║  User: ${user.padEnd(50)}║`);
  console.log(`║  Created: ${license.created.substring(0, 19).padEnd(47)}║`);
  console.log(`║  Expires: ${license.expires.substring(0, 19).padEnd(47)}║`);
  console.log('║                                                            ║');
  console.log('╠════════════════════════════════════════════════════════════╣');
  console.log('║  LICENSE KEY:                                              ║');
  console.log(`║  ${license.key.padEnd(55)}║`);
  console.log('║                                                            ║');
  console.log('╚════════════════════════════════════════════════════════════╝\n');
  
  // Save to licenses directory
  const licensesDir = path.join(__dirname, 'licenses');
  if (!fs.existsSync(licensesDir)) fs.mkdirSync(licensesDir, { recursive: true });
  fs.writeFileSync(path.join(licensesDir, `${user}_${Date.now()}.json`), JSON.stringify(license, null, 2));
  
  process.exit(0);
}

module.exports = {
  generateLicenseKey,
  verifyLicenseKey,
  checkLicense,
  activateLicense,
  saveLicense
};
